pkg/broker/handler/pool/retry/pool.go

```diff
@@ -0,0 +1,139 @@
+
+	"github.com/cloudevents/sdk-go/v2/protocol/pubsub"
+	"github.com/google/knative-gcp/pkg/broker/config"
+	"github.com/google/knative-gcp/pkg/broker/handler"
+	handlerctx "github.com/google/knative-gcp/pkg/broker/handler/context"
+	"github.com/google/knative-gcp/pkg/broker/handler/pool"
+	"github.com/google/knative-gcp/pkg/broker/handler/processors"
+	"github.com/google/knative-gcp/pkg/broker/handler/processors/deliver"
+)
+
+// TODO Retry and Fanout are using similar SyncPool struct, needing an interface to reduce redundancies.
+// SyncPool is the sync pool for retry handlers.
+// For each trigger in the config, it will attempt to create a handler.
+// It will also stop/delete the handler if the corresponding trigger is deleted
+// in the config.
+type SyncPool struct {
+	options *pool.Options
+	targets config.ReadonlyTargets
+	pool    sync.Map
+}
+
+// StartSyncPool starts the sync pool.
+func StartSyncPool(ctx context.Context, targets config.ReadonlyTargets, opts ...pool.Option) (*SyncPool, error) {
+	p := &SyncPool{
+		targets: targets,
+		options: pool.NewOptions(opts...),
+	}
+	if err := p.syncOnce(ctx); err != nil {
+		return nil, err
+	}
+	if p.options.SyncSignal != nil {
+		go p.watch(ctx)
+	}
+	return p, nil
+}
+
+func (p *SyncPool) watch(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-p.options.SyncSignal:
+			if err := p.syncOnce(ctx); err != nil {
+				logging.FromContext(ctx).Error("failed to sync handlers pool on watch signal", zap.Error(err))
+			}
+		}
+	}
+}
+
+func (p *SyncPool) syncOnce(ctx context.Context) error {
+	var errs int
+
+	p.pool.Range(func(key, value interface{}) bool {
+		// Each target represents a trigger.
+		if _, ok := p.targets.GetTargetByKey(key.(string)); !ok {
+			value.(*handler.Handler).Stop()
+			p.pool.Delete(key)
+		}
+		return true
+	})
+
+	p.targets.RangeAllTargets(func(t *config.Target) bool {
+		tk := config.TriggerKey(t.Namespace, t.Broker, t.Name)
+
+		// There is already a handler for the trigger, skip.
+		if _, ok := p.pool.Load(tk); ok {
+			return true
+		}
+
+		projectIDOpt := pubsub.WithProjectIDFromDefaultEnv()
+		if p.options.ProjectID != "" {
+			projectIDOpt = pubsub.WithProjectID(p.options.ProjectID)
+		}
+		ps, err := pubsub.New(ctx,
+			projectIDOpt,
+			pubsub.WithTopicID(t.RetryQueue.Topic),
+			pubsub.WithSubscriptionID(t.RetryQueue.Subscription),
+		)
+		if err != nil {
+			logging.FromContext(ctx).Error("failed to create pubsub protocol", zap.String("trigger", t.Name), zap.Error(err))
+			errs++
+			return true
+		}
+
+		h := &handler.Handler{
+			Timeout:      p.options.TimeoutPerEvent,
+			PubsubEvents: ps,
+			Processor: processors.ChainProcessors(
+				// TODO filter processor may be added in the future, but need more discussion for that.
+				&deliver.Processor{Requester: p.options.EventRequester},
+			),
+		}
+		// Start the handler with target in context.
+		h.Start(handlerctx.WithTarget(ctx, t), func(err error) {
```

Comment:
The deliver processor also needs the broker for reply: https://github.com/google/knative-gcp/blob/master/pkg/broker/handler/processors/deliver/processor.go#L44

Reply:
This also reminds me that we would need a test to test reply with pool

Reply:
I'll have second PR for adding a semi-e2e for retry